### ISSUE TITLE
Make restore AZ aware

### ISSUE THREAD

**Issue by Michal-Leszczynski** [scylladb/scylla-manager#4039]
During restore improvement meetings, it was mentioned that making SM AZ aware could speed up the restore process.
We should experiment with that and see the results.

**Comment by Michal-Leszczynski** [comment]
Unfortunately, I don't have a clear idea on how to safely use AZ information in SM restore.
@avikivity could you explain the idea behind it?

**Comment by Michal-Leszczynski** [comment]
cc: @karol-kokoszka @mykaul @tzach

**Comment by avikivity** [comment]
If datacenter.RF == count(datacenter.racks), then each rack gets one replica. Typical example is RF=3 and nr_racks=3.

If this holds, you can take a rack's backup and copy it to just one restored cluster rack, with nodetool refresh --load-and-stream --keep-rack (doesn't exist yet). This reduces the number of receivers from 3 to 1, and significantly reduces the compaction load.

**Comment by Michal-Leszczynski** [comment]
> This reduces the number of receivers from 3 to 1, and significantly reduces the compaction load.

The reduction in receivers is already achieved with --primary-replica-only, but I guess that streaming withing the same rack should be faster. 

Perhaps this would also speed up the post-restore repair, as (depending on data consistency during backup) less data would need to be transferred between the nodes during the repair.

**Comment by bhalevy** [comment]
Cc @regevran

**Comment by regevran** [comment]
This should be a scylladb issue, but as an optimization, not for the general case.

**Comment by karol-kokoszka** [comment]
The scenario where the DC replication factor equals to number of racks in the given datacenter is something what would require from Scylla Manager to know the mapping of (source rack -> destination rack). 
Then, scylla manager could assure that data from a single rack is restored (downloaded + call for l&s) only by nodes from the expected rack (assuming that `--keep-rack` flag is implemented in core), because l&s would always choose the node from the same rack.
Repair is not needed, because the whole replica is restored then (what is guaranteed by `--keep-rack` flag and the fact that RF == number of racks).
Right now, l&s is duplicating the work, because `--primary-replica-only` flag causes the data of the same replica to always be streamed to the same node. What eventually creates a need for the post-restore repair.

**Just to clarify, we assume that this optimization works only when the restore is done on exactly the same topology.**

Right now, backup manifest doesn't contain information about the rack.
The RF can be picked up from the dumped schema. It requires to parse the `CREATE KEYSPACE...` string.

We can extend the manifest, so that the rack information is included into the manifest, but it means that old backups won't have it.
So, this optimization would be applied only to new backups.

We won't include into 3.4.

And the prerequisite is to have the flag `---keep-rack` is implemented on core side.

**Comment by karol-kokoszka** [comment]
> If datacenter.RF == count(datacenter.racks), then each rack gets one replica. Typical example is RF=3 and nr_racks=3.

> If this holds, you can take a rack's backup and copy it to just one restored cluster rack, with nodetool refresh --load-and-stream --keep-rack (doesn't exist yet). This reduces the number of receivers from 3 to 1, and significantly reduces the compaction load.

It’s a bit counterintuitive to combine `--primary-replica-only` with `--keep-rack`. The primary replica is unique for a given partition key, while `--keep-rack` means that data is streamed to replica belonging to the same rack that the caller belongs to. Perhaps, instead of combining `--keep-rack` and `--primary-replica-only`, there should be a single flag that defines the desired behavior. For example, using `--keep-rack` without `--primary-replica-only` might make more sense. It should be possible to use either `--keep-rack` or `--primary-replica-only` independently.

**Comment by bhalevy** [comment]
> This should be a scylladb issue, but as an optimization, not for the general case.

@regevran please open issue for implementing nodetool refresh --load-and-stream --keep-rack
Cc @denesb

**Comment by bhalevy** [comment]
> Currently, the backup manifest does not include rack information.

FWIW, this can be retrieved from system.local or system.peers if they are backed up.
the scylla sstable tool cab be used to dump their contents in json format.

**Comment by regevran** [comment]
> please open issue for implementing nodetool refresh --load-and-stream --keep-rack

[#4798](https://github.com/scylladb/scylla-enterprise/issues/4798)

**Comment by karol-kokoszka** [comment]
@regevran @bhalevy 

Before applying this change to the Scylla Manager restore using native restore from Scylla for tablets.

I can imagine the situation, where SM hits a node from rack that doesn't own any data of this particular keyspace if the number of racks is greater than RF. What then ? Will the SSTables be drop ? 
Will the API respond - "no keyspace replica in this rack - try A, B, C" ? 

Is it common that customers keep the RF = rack number ?

If the above is false, Scylla Manager will need to validate the pre-requisites before calling for this variation of the restore.

**Comment by bhalevy** [comment]
@karol-kokoszka isn't SM restoring sstables that were on the rack when backed up?

As ai discussed with @mykaul we should require the restored cluster to run the same software version as tye backed up cluster, and they should allocate the same racks if rf < nr_racks.

The common case is that rf = nr_racks indeed.

Also, @tgrabiec, @avikivity I think we should commit to keeping the rack names in the schema rf options and not just rf count per dc, so that the mapping will be stable and can specified in create or alter table for restore.

**Comment by karol-kokoszka** [comment]
> @karol-kokoszka isn't SM restoring sstables that were on the rack when backed up?

Yes, it is, but is it enough ? Can we assume that the tablets are distributed the same way as in the backup ? 
Is the cluster expected to keep the topology ? 

> As ai discussed with @mykaul we should require the restored cluster to run the same software version as tye backed up cluster, and they should allocate the same racks if rf < nr_racks.

If the rack allocation is the same, if the tablets are in exactly the same racks as in the snapshot - OK. But I'm not sure if that's true in the context of this issue.

> The common case is that rf = nr_racks indeed.

> Also, @tgrabiec, @avikivity I think we should commit to keeping the rack names in the schema rf options and not just rf count per dc, so that the mapping will be stable and can specified in create or alter table for restore.

I think, it's the user who bootstraps the cluster to assure that mapping source-rack -> target-rack is OK and needs to provide the mapping to Scylla Manager (at least this is what we do for vnodes 1-1 restore, source-host-id -> dest-host-iid mapping is passed as a part of restore request).
But I'm not sure if rack mapping is enough. It's more about tablets distribution.

**Comment by mykaul** [comment]
> Yes, it is, but is it enough ? Can we assume that the tablets are distributed the same way as in the backup ?
> Is the cluster expected to keep the topology ?

I assume the AZs setup is similar (but may not be 1:1 - this is not a 1:1, so you may have zone A, B, C and now zone A, B, D)
Cluster topology (no. of AZs) should be the same as before
Per AZ, node distribution will be the same (and you restore per AZ). Intra-AZ - not sure. You may have a different count of nodes and shards in theory. I think the load balancer will be able to balance shards later if needed.

**Comment by regevran** [comment]
>  I think the load balancer will be able to balance shards later if needed.

This is true but it has a price - unbalanced cluster may experience latencies.

**Comment by regevran** [comment]
> I understand that we want to keep the same amount of nodes and shards per node as well ?

In the 1:1 restore scenario only.

**Comment by mykaul** [comment]
> > I think the load balancer will be able to balance shards later if needed.
> 
> This is true but it has a price - unbalanced cluster may experience latencies.

Don't let perfect be the enemy of good.

I expect 1:1 to be a much more popular use case and rack awareness is good and important , but less critical.
The real question though: what if we can restore much faster, by using a 2X sized cluster, then scale-in to the original size? This is an interesting use case. It's 1:1, the Next Generation(TM). 
In this case, you do have the same topology, perhaps the same number of nodes, but different shard count, which allows you to download in parallel from more shards, etc.

**Comment by pdbossman** [comment]
> The scenario where the DC replication factor equals the number of racks in a given datacenter requires Scylla Manager to understand the mapping between the source rack and the destination rack. In this case, Scylla Manager could ensure that data from a single rack is restored (downloaded and then subjected to l&s) only by nodes from the corresponding rack, assuming that the --keep-rack flag is implemented in the core. This is because the l&s process would always select a node from the same rack.
> 
> A repair would not be necessary because the fact that entire replica is restored, is guaranteed by the --keep-rack flag and the fact that the RF equals the number of racks. Currently, l&s duplicates work because the --primary-replica-only flag causes the same replica's data to always be streamed to the same node, which eventually necessitates a post-restore repair.
> 
> To clarify, we assume this optimization applies only when the restore occurs on an identical topology.
> 
> Currently, the backup manifest does not include rack information. The RF can be determined from the dumped schema by parsing the CREATE KEYSPACE... string.
> 
> We could extend the manifest to include rack information, but this would mean that older backups would not contain this information. As a result, the optimization would only apply to new backups.
> 
> We will not include this in version 3.4. The prerequisite for this optimization is that the --keep-rack flag is implemented on the core side.

Agree with @mykaul about not trying to be perfect... from a practical perspective if/when we test rack awareness please consider:
If a region dies, we're restoring to a different region, so we may have different DC names and different rack (names).  But the type, number of nodes, disk per node (GCP) should be identical.

**Comment by regevran** [comment]
Restore can be done file-based: the entire SSTable is being restored as-is.
Or it can be mutation based: SSTable is being read mutation by mutation - effectively creating a new SStable.
The fastest restore is the one that file-restore ALL sstables.
For this to happen - all SStables in the backup must fit into the tablet map of the target. Since we are restoring the tablet map as well, we can set the number of tablet to perfectly fit the stored SSTables. 
However, this tablet map that we build depending on the stored SSTables does not necessarily fit best the target cluster.
For example: if we doubled the cluster size, and use the same tablet map - we use half of the available resources, the cluster is not balanced. On the other hand, if we double the tablet map then we'll not be able to restore using file-based, because we need to split the stored SSTables to fit into the new tablet map. 

The trade off is between unbalanced-but-fast-file-based-restore to balanced-but-slower-mutation-based-restore.
The truth and best approach is probably not on one of those ends, but somewhere in between.
E.g. creating a tablet map for the target that maximizes the number of tablets but allows only up to some percentage of mutation based restore.

And last thing: 1:1 restore offers the perfect match - the tablet size is perfect for the cluster and for the stored SSTables.

**Comment by karol-kokoszka** [comment]
> For this to happen - all SStables in the backup must fit into the tablet map of the target. **Since we are restoring the tablet map as well,** we can set the number of tablet to perfectly fit the stored SSTables.

I didn't know you are restoring the tablets map.
Please correct me if I'm wrong, but you are referring to this system table that stores the mapping and which cannot be restored at the moment ?

In my understanding, if the tablets map is restored, we are good to go.

But the tablets balancing must be disabled.

**Comment by regevran** [comment]
We are not restoring the tablet map - this is left to Scylla.
We just give a hint, we set `min_tablet_count`.
Setting it correctly will make get the best from user perspective: fastest download _and_ good service.
If we prioritize fast download we pay with slow service (unbalanced cluster) after the restore. If we prioritize good service (balanced cluster) we pay with slow download.

**Comment by pdbossman** [comment]
> We are not restoring the tablet map - this is left to Scylla. We just give a hint, we set `min_tablet_count`. Setting it correctly will make get the best from user perspective: fastest download _and_ good service. If we prioritize fast download we pay with slow service (unbalanced cluster) after the restore. If we prioritize good service (balanced cluster) we pay with slow download.

Speaking as a former enterprise end user and SME in this space.

The purpose of fast restore - and I apologize for the all caps here but - THE CUSTOMER IS DOWN.  100% down.

Whether they buy scylla or not, some customers have to prove meeting RTO - recovery time objective.  When I measured this as an end user, it was time from outage start to app is back up serving.

**_Shorten the time to them being able to use scylla._**

**Comment by regevran** [comment]
When you say "Shorten the time to them being able to use scylla." -
what does that mean to use scylla?

Are you ok with high p99? If yes, for how long?

—

**Comment by pdbossman** [comment]
> When you say "Shorten the time to them being able to use scylla." -
> what does that mean to use scylla?
> 
> Are you ok with high p99? If yes, for how long?
> 
> —
> […](#)

@regevran until the data is available, the p99 is infinity, right?
After the data is available, if we're tablets, we can start auto-scaling out to address p99?

**Comment by regevran** [comment]
@pdbossman, I am asking because I really would like to know.
For me restore + reasonable-response-times _together_ are the end of the restore process.

I had this idea of "instant restore" - in this mode we start serving immediately from S3 directly - no download, nothing. Then we download in the background, with horrible latency meanwhile. This approach is the fastest I know that brings the cluster into service. But a bad service. 

At the other end we have carefully restoring mutation by mutation the entire cluster until it is fully restored with all data in the exact place as it should be, serving at high speeds.

I guess that both ends are not good enough - we are searching for the point in the middle that we optimize both aspects.

**Comment by tzach** [comment]
> I had this idea of "instant restore" - in this mode we start serving immediately from S3 directly - no download, nothing. Then we download in the background, with horrible latency meanwhile. This approach is the fastest I know that brings the cluster into service. But a bad service.

This is useless for most ScyllaDB use cases when low latency is assumed.

**Comment by pdbossman** [comment]
> [@pdbossman](https://github.com/pdbossman), I am asking because I really would like to know. For me restore + reasonable-response-times _together_ are the end of the restore process.
> 
> I had this idea of "instant restore" - in this mode we start serving immediately from S3 directly - no download, nothing. Then we download in the background, with horrible latency meanwhile. This approach is the fastest I know that brings the cluster into service. But a bad service.
> 
> At the other end we have carefully restoring mutation by mutation the entire cluster until it is fully restored with all data in the exact place as it should be, serving at high speeds.
> 
> I guess that both ends are not good enough - we are searching for the point in the middle that we optimize both aspects.

1-1 restore means we are back as they were, right?  We will have to warm up, but I am struggling to understand why we think performance will be bad on topology where they were.

**Comment by tzach** [comment]
> 1-1 restore means we are back as they were, right? We will have to warm up, but I am struggling to understand why we think performance will be bad on topology where they were.

It would not be bad, but restoring to the x2 cluster means doubling the download speed.

**Comment by pdbossman** [comment]
> > 1-1 restore means we are back as they were, right? We will have to warm up, but I am struggling to understand why we think performance will be bad on topology where they were.
> 
> It would not be bad, but restoring to the x2 cluster means doubling the download speed.

No one will complain if you can reliably get 1 : 2 or 1 : 3 restore to run in 50% (or 33%) the time as 1:1.

**Comment by regevran** [comment]
On Mon, Apr 28, 2025, 20:39 Tzach Livyatan ***@***.***> wrote:

> *tzach* left a comment (scylladb/scylla-manager#4039)
> <https://github.com/scylladb/scylla-manager/issues/4039#issuecomment-2836003157>
>
> 1-1 restore means we are back as they were, right? We will have to warm
> up, but I am struggling to understand why we think performance will be bad
> on topology where they were.
>
> It would not be bad, but restoring to the x2 cluster means doubling the
> download speed.
>

Yes, but once you done downloading and start serving, you may be
potentially paying the price of high latencies because of unbalanced
cluster, that tries to balance itself.

**Comment by pdbossman** [comment]
> On Mon, Apr 28, 2025, 20:39 Tzach Livyatan ***@***.***> wrote:
>  *tzach* left a comment ([scylladb/scylla-manager#4039](https://github.com/scylladb/scylla-manager/issues/4039))
>  <[#4039 (comment)](https://github.com/scylladb/scylla-manager/issues/4039#issuecomment-2836003157)>
> 
>  1-1 restore means we are back as they were, right? We will have to warm
>  up, but I am struggling to understand why we think performance will be bad
>  on topology where they were.
> 
>  It would not be bad, but restoring to the x2 cluster means doubling the
>  download speed.
> Yes, but once you done downloading and start serving, you may be
> potentially paying the price of high latencies because of unbalanced
> cluster, that tries to balance itself.
> […](#)

Wouldn't the original cluster have to have been unbalanced for the new target to be imbalanced?

**Comment by pdbossman** [comment]
1:1 is pretty much like a clone of the cluster at the time backup was taken.  Whatever the balance was on the source, it is replicated on the target, right?

**Comment by regevran** [comment]
On Mon, Apr 28, 2025, 20:23 Patrick Bossman ***@***.***>
wrote:

> *pdbossman* left a comment (scylladb/scylla-manager#4039)
> <https://github.com/scylladb/scylla-manager/issues/4039#issuecomment-2835962207>
>
> @pdbossman <https://github.com/pdbossman>, I am asking because I really
> would like to know. For me restore + reasonable-response-times *together*
> are the end of the restore process.
>
> I had this idea of "instant restore" - in this mode we start serving
> immediately from S3 directly - no download, nothing. Then we download in
> the background, with horrible latency meanwhile. This approach is the
> fastest I know that brings the cluster into service. But a bad service.
>
> At the other end we have carefully restoring mutation by mutation the
> entire cluster until it is fully restored with all data in the exact place
> as it should be, serving at high speeds.
>
> I guess that both ends are not good enough - we are searching for the
> point in the middle that we optimize both aspects.
>
> 1-1 restore means we are back as they were, right? We will have to warm
> up, but I am struggling to understand why we think performance will be bad
> on topology where they were.
>
1:1 restore is the perfect restore from all perspectives, the data is being
downloaded in the fastest possible manner _and_ it fits into the best
tablet map exists (or vnodes).

But 1:1 restore is a special case of the general restore case.
However, since it is common enough (maybe THE common case) we are all good
once it is implemented.

**Comment by regevran** [comment]
On Mon, Apr 28, 2025, 20:11 Tzach Livyatan ***@***.***> wrote:

> *tzach* left a comment (scylladb/scylla-manager#4039)
> <https://github.com/scylladb/scylla-manager/issues/4039#issuecomment-2835929453>
>
> I had this idea of "instant restore" - in this mode we start serving
> immediately from S3 directly - no download, nothing. Then we download in
> the background, with horrible latency meanwhile. This approach is the
> fastest I know that brings the cluster into service. But a bad service.
>
> This is useless for most ScyllaDB use cases when low latency is assumed.
>
Yes. It was explained to illustrate both ends of data-readiness.

**Comment by regevran** [comment]
On Mon, Apr 28, 2025, 22:45 Patrick Bossman ***@***.***>
wrote:

> *pdbossman* left a comment (scylladb/scylla-manager#4039)
> <https://github.com/scylladb/scylla-manager/issues/4039#issuecomment-2836336878>
>
> 1:1 is pretty much like a clone of the cluster at the time backup was
> taken. Whatever the balance was on the source, it is replicated on the
> target, right?
>

Yes. See my other response. 1:1 is the best. Doubling the cluster to get
double download throughout is not 1:1 and therefore might not be the best
in terms of ready to serve restore.

**Comment by regevran** [comment]
On Mon, Apr 28, 2025, 22:45 Patrick Bossman ***@***.***>
wrote:

> *pdbossman* left a comment (scylladb/scylla-manager#4039)
> <https://github.com/scylladb/scylla-manager/issues/4039#issuecomment-2836336878>
>
> 1:1 is pretty much like a clone of the cluster at the time backup was
> taken. Whatever the balance was on the source, it is replicated on the
> target, right?
>
With 1:1 restore yes. In the general case no. The assumption is that at the
time of backup the cluster was balanced. Retuning this balance to a cluster
with a different topology is hard, and requires compromising download
speed.

**Comment by tgrabiec** [comment]
> On Mon, Apr 28, 2025, 20:39 Tzach Livyatan ***@***.***> wrote:
>  *tzach* left a comment ([scylladb/scylla-manager#4039](https://github.com/scylladb/scylla-manager/issues/4039))
>  <[#4039 (comment)](https://github.com/scylladb/scylla-manager/issues/4039#issuecomment-2836003157)>
> 
>  1-1 restore means we are back as they were, right? We will have to warm
>  up, but I am struggling to understand why we think performance will be bad
>  on topology where they were.
> 
>  It would not be bad, but restoring to the x2 cluster means doubling the
>  download speed.
> Yes, but once you done downloading and start serving, you may be
> potentially paying the price of high latencies because of unbalanced
> cluster, that tries to balance itself.
> […](#)

But you have twice the processing capacity. If before backup we had 10 tablets/shard so up to 10% imbalance in shard utilization. Assuming the worst case scenario of 90% and 100% utilization, with 2x cluster we will have 5 tablets/shard so up to 20% imbalance of utilization, but the actual utilization, assuming the same workload, will be 40% and 50%. You will get much better p99 despite greater imbalance, due to smaller utilization.

**Comment by regevran** [comment]
On Mon, Apr 28, 2025, 22:59 Tomasz Grabiec ***@***.***> wrote:

> *tgrabiec* left a comment (scylladb/scylla-manager#4039)
> <https://github.com/scylladb/scylla-manager/issues/4039#issuecomment-2836372878>
>
> On Mon, Apr 28, 2025, 20:39 Tzach Livyatan *@*.***> wrote:
> *tzach* left a comment (scylladb/scylla-manager#4039
> <https://github.com/scylladb/scylla-manager/issues/4039>)
> <#4039 (comment)
> <https://github.com/scylladb/scylla-manager/issues/4039#issuecomment-2836003157>
> >
>
> 1-1 restore means we are back as they were, right? We will have to warm
> up, but I am struggling to understand why we think performance will be bad
> on topology where they were.
>
> It would not be bad, but restoring to the x2 cluster means doubling the
> download speed.
> Yes, but once you done downloading and start serving, you may be
> potentially paying the price of high latencies because of unbalanced
> cluster, that tries to balance itself.
> … <#m_-8160108620026515057_>
>
> But you have twice the processing capacity. If before backup we had 10
> tablets/shard so up to 10% imbalance in shard utilization. Assuming the
> worst case scenario of 90% and 100% utilization, with 2x cluster we will
> have 5 tablets/shard so up to 20% imbalance of utilization, but the actual
> utilization, assuming the same workload, will be 40% and 50%. You will get
> much better p99 despite greater imbalance, due to smaller utilization.
>

if you are able to keep the same tablet map as in the backup, then I guess
you are correct. But the assumption was that doubling the cluster allows
you to double the tablet map and then fallback to mutation restore which is
slower than file restore.

**Comment by bhalevy** [comment]
> On Mon, Apr 28, 2025, 22:59 Tomasz Grabiec ***@***.***> wrote:
>  *tgrabiec* left a comment ([scylladb/scylla-manager#4039](https://github.com/scylladb/scylla-manager/issues/4039))
>  <[#4039 (comment)](https://github.com/scylladb/scylla-manager/issues/4039#issuecomment-2836372878)>
> 
>  On Mon, Apr 28, 2025, 20:39 Tzach Livyatan *@*.***> wrote:
>  *tzach* left a comment ([scylladb/scylla-manager#4039](https://github.com/scylladb/scylla-manager/issues/4039)
>  <[#4039](https://github.com/scylladb/scylla-manager/issues/4039)>)
>  <[#4039](https://github.com/scylladb/scylla-manager/issues/4039) (comment)
>  <[#4039 (comment)](https://github.com/scylladb/scylla-manager/issues/4039#issuecomment-2836003157)>
>  >
> 
>  1-1 restore means we are back as they were, right? We will have to warm
>  up, but I am struggling to understand why we think performance will be bad
>  on topology where they were.
> 
>  It would not be bad, but restoring to the x2 cluster means doubling the
>  download speed.
>  Yes, but once you done downloading and start serving, you may be
>  potentially paying the price of high latencies because of unbalanced
>  cluster, that tries to balance itself.
>  … <#m_-8160108620026515057_>
> 
>  But you have twice the processing capacity. If before backup we had 10
>  tablets/shard so up to 10% imbalance in shard utilization. Assuming the
>  worst case scenario of 90% and 100% utilization, with 2x cluster we will
>  have 5 tablets/shard so up to 20% imbalance of utilization, but the actual
>  utilization, assuming the same workload, will be 40% and 50%. You will get
>  much better p99 despite greater imbalance, due to smaller utilization.
> if you are able to keep the same tablet map as in the backup, then I guess
> you are correct. But the assumption was that doubling the cluster allows
> you to double the tablet map and then fallback to mutation restore which is
> slower than file restore.
> […](#)

Not sure where this assumption is coming from.
The point in artificially doubling (or more) of the cluster is to reduce restore time by providing more bandwidth to S3.
It is true that preserving the number of tablets as it was at backup time may lead to under-utilization even to the point that a table won't be served by all shards, but this would be true also if the same table was served by the artificially large cluster unrelated to backup and restore, since we aim at a given target tablet size (5 GB by default).
We may inflate tablet count initially, when a table is created, to reduce cross-shard imbalance, but eventually the tablet count will be adjusted dynamically as more tables are added and data is loaded to the table(s).

Bottom line is that for restore purposes, we may choose to oversize the cluster to reduce restore time, and live with under-utilization after backup until the cluster is scaled-in to reach the desired utilization level (e.g. 80-90%).

**Comment by Michal-Leszczynski** [comment]
> yes, using scope=dc tell restore or load-and-stream to send the mutation to the primary replica in the node's DC for each token that is restored. This will help avoid cross-zone communication as much as possible to reduce cost and increase performance.

Is this always safe in the context of restore (just double checking here)?
If so, I will create an issue and fix it shortly, as this is a minimal change from SM POV.

> As for the following repair stage. we cannot avoid it, but we can repair each DC independently (e.g. with /storage_service/tablets/repair using the dcs_filter option) and then run a global repair

From what I understood (@Deexie please correct me if I'm wrong), we can't repair multiple dcs in parallel. So we would end up repairing each dc sequentially before running repair targeting all dcs. Because of that, I'm not sure if the saving of traffic/money is worth the sequential delay.
It also sounds like a enhancement that should be performed by scylla internally when multi dc repair is scheduled, is that right?
